common.rkt

```diff
@@ -140,23 +140,35 @@
 (define (reify tm st)
   (define index -1)
   (let ((results (let loop ((tm tm) (st st))
-              (define t (walk tm (state-sub st)))
-              (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
-                    ((var? t)  (set! index (+ 1 index))
-                               (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
-                    (else      st)))))
+                   (define t (walk tm (state-sub st)))
+                   (cond ((pair? t) (loop (cdr t) (loop (car t) st)))
+                         ((var? t)  (set! index (+ 1 index))
+                                    (state (extend-sub t (reified-index index) (state-sub st)) (state-diseq st) (state-types st)))
+                         (else      st)))))
     (let* ((walked-sub (walk* tm results))
+           (diseq (sort (state-diseq st) (lambda (x y) (< (length x) (length y)))))
+           (st (diseq-simplify (state (state-sub st) diseq (state-types st))))
+           (diseq (walk* (map pretty-diseq (state-diseq st)) results))
+           (diseq (filter fresh-var-diseq? diseq))
+           (diseq (if (null? diseq) '() (list (cons '=/= diseq))))
            (types (walk* (map pretty-types (state-types st)) results))
-           (diseq (map pretty-diseq (state-diseq st)))
-           (diseq (if (null? diseq) '() (list (walk* (cons '=/= diseq) results))))
+           (types (filter (lambda (t) (not (contains-fresh? (cdr t)))) types))
            (cxs (append types diseq)))
       (if (null? cxs)
           walked-sub
-          (Ans walked-sub cxs)))))
+          (Ans walked-sub (sort cxs term<?))))))
 
 (define (reify/initial-var st)
   (reify initial-var st))
 
+(define (term<? u v)
+  (and (not (eqv? (car u) '=/=)) (eqv? (car v) '=/=)))
```

Comment:
Is this a placeholder?  In general you'll need to sort within all of the usual types of terms, as well as giving an order to the types themselves:

Here is one possible inter-type order.  And then intra-type ordering is defined by a type-specific predicate in each case:

```
()
#f
#t
numbers ; <
strings ; string<?
symbols ; symbol<?
pairs   ; recursively and lexicographically sort car, then cdr (i.e., if cars are unequal, you're done, otherwise try the cdr)
anything-else-that-isn't-one-of-the-types-above  ; don't worry about intra-type sorting of these, just make sure they appear at the end
```

Reply:
I'm a little confused on how we go about recursively sorting the car and cdr when they are pairs.

`term<?` is supposed to return `#t` or `#f`, so it shouldn't mutate the car or cdr (unless I've misunderstood how `sort` works), so when would this recursive sort take place?

Reply:
Oops.  By recursively sort I meant that `term<?` should recursively compare the components as necessary.  Sorry about that.